File: 프로그래머스/lv3/49191. 순위/순위.py
from collections import deque

def solution(n, results):
    
    def bfs(start, graph, individual_nodes):
        queue = deque([]); queue.append(start)
        visit = [False] * (n+1); visit[start] = True
        
        while queue:
            curr = queue.popleft()
            individual_nodes[curr].add(curr)
            
            if curr not in graph: continue
            
            for element in graph[curr]:
                if not visit[element]:
                    visit[element] = True
                    queue.append(element)
                    for c in individual_nodes[curr]:
                        individual_nodes[element].add(c)
        
    
    win_graph = dict(); lose_graph = dict()
    for result in results:
        w, l = result
        if w in win_graph: win_graph[w].append(l)
        else: win_graph[w] = [l]
        if l in lose_graph: lose_graph[l].append(w)
        else: lose_graph[l] = [w]
    
    win_nodes = dict(); lose_nodes = dict()
    for i in range(1, n+1):
        win_nodes[i] = set(); lose_nodes[i] = set()
    
    for i in range(1, n+1):
        bfs(i, win_graph, win_nodes)
        bfs(i, lose_graph, lose_nodes)
    
    answer = 0
    for i in range(1, n+1):
        if len(win_nodes[i]) + len(lose_nodes[i]) == n+1: answer += 1
    return answer

File: 프로그래머스/3/49191. 순위/순위.cpp
#include <string>
#include <vector>
#include <deque>
#include <map>
#include <set>
#include <iostream>

using namespace std;

int N;

void bfs(int start, map<int, vector<int>> &graph, map<int, set<int>> &nodes)
{
    vector<bool> visit(N+1, false); visit[start] = true;
    deque<int> queue; queue.push_back(start);
    
    while (!queue.empty())
    {
        int cur = queue.front(); queue.pop_front();
        nodes[cur].insert(cur);
        
        for (auto nex : graph[cur])
        {
            if (!visit[nex])
            {
                visit[nex] = true;
                queue.push_back(nex);
                
                for (auto it : nodes[cur])
                {
                    nodes[nex].insert(it);
                }
            }
        }
        
    }
    
}

int solution(int n, vector<vector<int>> results) {
    
    N = n;
    
    map<int, vector<int>> win_graph;
    map<int, vector<int>> lose_graph;
    for (auto result : results)
    {
        win_graph[result[0]].push_back(result[1]);
        lose_graph[result[1]].push_back(result[0]);
    }
    
    map<int, set<int>> win_nodes;   //bfs 탐색하면서 접근 가능한 모든 nodes들을 모음
    map<int, set<int>> lose_nodes;  //bfs 탐색하면서 접근 가능한 모든 nodes들을 모음
    for (int i = 1; i <= N; ++i)
    {
        bfs(i, win_graph, win_nodes);
        bfs(i, lose_graph, lose_nodes);
    }
    
    int ans = 0;
    for (int i = 1; i <= N; ++i)
    {
        // cout << win_nodes[i].size() << ", " << lose_nodes[i].size() << endl;
        if (win_nodes[i].size() + lose_nodes[i].size() == N+1)
        {
            ans++;
        }
    }
    
    return ans;
}